### ISSUE TITLE
Improve sql fusion / avoid nested selects / nested queries

### ISSUE THREAD

**Issue by szeiger** [slick/slick#489]
*[Migrated from Assembla ticket [290](https://www.assembla.com/spaces/typesafe-slick/tickets/290) - reported by @cvogt on 2013-07-12 18:04:53]*

There is an issue discussed here https://t.co/FJKz1STcvi that a simple select count(*) from table cant be generated using the lifted embedding. table.length generates three nested selects instead.

The issue can be reproduced using
sbt test-only -- +q -Dlog.qcomp=debug *CountTest*
and searching the output for "select count".

**Comment by cmacher** [comment]
Changing the explicit joins to implicit joins fixed the performance issues for me. Not ideal but a workaround I can live with for now.

**Comment by cmacher** [comment]
Would #1135 be a way around this issue or is this unrelated?

**Comment by szeiger** [comment]
Purely aliasing subqueries should be completely gone in 3.1. We'll need some additional reordering for unions (pushing projections down into both sides). For joins it's probably better not to do this. We can look at the details of additional optimizations once the rewrite is complete.

With the new back-end, monadic joins are always translated to explicit applicative joins first. When a Comprehension contains only inner/cross joins, the code generator uses implicit syntax (which is simpler), otherwise explicit (because mixing may not be supported).

Here's the code from a simple explicit join over 3 tables with join conditions and an additional projection in 3.0:

```
select s3.s50, s3.s54, s4.s58
from (
  select s5.s59 as s50, s6.s63 as s54
  from (
    select s74."id" as s59, s74."a" as s60, s74."b" as s61
    from "A_NEWFUSION" s74
  ) s5
  inner join (
    select s77."id" as s62, s77."a" as s63, s77."b" as s64
    from "A_NEWFUSION" s77
  ) s6
  on s5.s59 = s6.s62
) s3
inner join (
  select s80."id" as s56, s80."b" as s58
  from "A_NEWFUSION" s80
) s4
on s3.s50 = s4.s56
```

And the same in my current working branch:

```
select s5."id", s6."a", s4."b"
from "A_NEWFUSION" s5, "A_NEWFUSION" s6, "A_NEWFUSION" s4
where (s5."id" = s6."id") and (s5."id" = s4."id")
```

**Comment by ijuma** [comment]
Nice improvement. :)

**Comment by nremond** [comment]
Great work !

**Comment by loverdos** [comment]
Sweet!

**Comment by godenji** [comment]
@szeiger what will the generated sql look like when you can't desugar to implicit join syntax? i.e. with outer joins, are you anticipating similar (or close to) generated sql in 3.1?

If so, that would be a huge performance win for MySQL, and probably allow for improved query plans in other engines as well.

Tackling groupBy is the final frontier ;-)

**Comment by szeiger** [comment]
@godenji The final decision between join syntaxes is at the code generator level. The shape of the query is exactly the same. Without optimizing simple joins to implicit syntax it would look like this (same for outer joins):

```
select s5."id", s6."a", s4."b"
from "A_NEWFUSION" s5
join "A_NEWFUSION" s6 on s5."id" = s6."id"
join "A_NEWFUSION" s4 on s5."id" = s4."id"
```

(Maybe nested with parentheses or in a different order; I just edited this one manually)

GroupBy was rather simple to add to the new back-end. I thought it would be challenging but turned out to be simple. Non-grouping aggregations on the other hand are still a problem that I'm working on.

**Comment by godenji** [comment]
@szeiger  
> Without optimizing simple joins to implicit syntax it would look like this (same for outer joins) ...

Incredible, 3.1 should be well received across the board for that improvement alone.

> Non-grouping aggregations on the other hand are still a problem that I'm working on

Do you mean handling `select max(foo) from bar`?  Really, aside from the comprehensive for comprehensions proposal getting shelved\*, the main issue I have with the current groupBy implementation is the requirement to flatten results. 

From SQL99 onward it is legal to return non-aggregate columns in a group by query provided that the group by clause contain at least one column from each table for which a functional dependency exists. Basically, given a query like: 

`select u.*, o.* from user u join orders o on u.id = o.userId group by o.userId, u.id`

you are allowed to return all columns from each table since the dependency/uniqueness constraint has been satisfied in the group by. Obviously this is not possible in current Slick as the result set needs to be flattened via min, max, etc. The workaround, grouping client-side, is unfortunate. Perhaps this is what you meant by "Non-grouping aggregations"?

\* though your `groupMap` attempt would be an improvement, perhaps that will type check in future (I tried, no luck on 2.11.6)

**Comment by szeiger** [comment]
Fixed in 3.1.
